### DISCUSSION TITLE
RAO of an elastic structure

### DISCUSSION THREAD

**Discussion by hojjatanvari** [capytaine/capytaine#632] [Q&A]
Dear All,

We know the simple code for defining a rigid body is something like this:

`cylinder_mesh = load_MAR("Cylinder1.dat", name="cylinder")`
`cylinder_body = cpt.FloatingBody(mesh=cylinder_mesh,  dofs=cpt.rigid_body_dofs(rotation_center=(0.0, 0.0, 0.0)), center_of_mass=(0.0, 0.0, 0.0))
`

I calculated RAO and other parameters of the rigid structure in six degrees of freedom. The problem is that I'd like to do this process for an _elastic structure_. I did a modal analysis in ABAQUS for the structure. Assuming the first 10 mode shape data is available, how can I import this data in the Capytaine code so that it considers the elastic structure in calculations?

Thanks.

**Comment by mancellin** [comment]
There is this example https://github.com/capytaine/capytaine/blob/37a2bbe29977903d28fcdf2785c2859f638874d5/docs/user_manual/examples/elasticity_of_beam.py (which has been mistakenly forgotten from the documentation cookbook).

**Reply by hojjatanvari** [reply]
Hi mancellin,
Thanks for sharing that example with me.
The structure I am working on is a series of cylinders and I want to have all as a single body and calculate the RAO of this body. 
This is why I want to import mode shape data of the whole structure and then find the body's RAO. 
I've got stuck in this step because calculating mode shapes for each cylinder (like in that example) is not correct. I created the mesh files of the structure using Capytaine commands but not sure how to assign my mode data to these meshes. Do you have any idea how to do this process?
Thanks in advance for your help.

**Reply by mancellin** [reply]
Assuming you have a Python function `deformation_mode(i, x, y, z)` that read your Abaqus data for the deformation at position $(x, y, z)$ for mode $i$ and returns a vector $(dx, dy, dz)$, then something like the following should work. I have no idea how the Abaqus data are structured so I can't really help with that.

```python
mesh = ...  # No need to be connex, could be the composition of several shapes

dofs = {}
for i in range(nb_modes):
    dofs[f"mode_{i}"] = np.array([deformation_mode(i, x, y, z) for (x, y, z) in mesh.faces_centers])

body = cpt.FloatingBody(mesh=mesh, dofs=dofs)

# Check
body.animate({"mode_1": 0.1}, loop_duration=1.0).run(top_light_intensity=0.0)
```

**Reply by hojjatanvari** [reply]
Thanks so much, mancellin.
I got your point. What I did was importing mesh files(.stl) from Fusion and mode shape data (.csv) from ABAQUS and then interpolating mode shape data on the mesh.
